File: python_workflow/tests/test_generation_subsyllabic.py
"""The subsyllabic (Wuggy-style) pseudoword generator."""
import pytest

from lexsync.generation import (_legal, bigram_counts, build_constituent_inventory,
                                build_lexdec_stimuli, generate_pseudowords_subsyllabic,
                                segment_subsyllabic)
from lexsync.querying import build_pool, load_lexicon


def test_segmentation_worked_examples():
    assert segment_subsyllabic("bridge") == [
        ("onset", "br"), ("nucleus", "i"), ("coda", "d"), ("onset", "g"), ("nucleus", "e")]
    assert segment_subsyllabic("planet") == [
        ("onset", "pl"), ("nucleus", "a"), ("onset", "n"), ("nucleus", "e"), ("coda", "t")]
    assert segment_subsyllabic("strength") == [
        ("onset", "str"), ("nucleus", "e"), ("coda", "ngth")]
    assert segment_subsyllabic("a") == [("nucleus", "a")]
    assert segment_subsyllabic("crwth") == []          # no vowel run -> empty (caller falls back)


def test_inventory_is_order_independent():
    ref = ["bridge", "planet", "strength", "brake", "grape", "plane", "clip", "drum"]
    assert build_constituent_inventory(ref) == build_constituent_inventory(list(reversed(ref)))


def test_generate_subsyllabic_legal_nonword_length_preserved(schema, en_lexicon_path):
    lex = load_lexicon(en_lexicon_path, schema, "english")
    ref = lex["word"].tolist()
    base = build_pool(lex, {"length": [4, 7]})["word"].tolist()[:30]
    gen = generate_pseudowords_subsyllabic(base, ref)
    lexset = set(ref)
    bg = bigram_counts(ref)
    for w, pw in zip(gen["base_word"], gen["pseudoword"]):
        assert len(pw) == len(w)             # length preserved exactly
        assert pw not in lexset              # a novel non-word
        assert _legal(pw, bg)                # every bigram attested
    assert gen["pseudoword"].nunique() == len(gen)     # all distinct


def test_build_lexdec_unknown_method_raises(schema, en_lexicon_path):
    lex = load_lexicon(en_lexicon_path, schema, "english")
    pool = build_pool(lex, {"length": [4, 6]})
    with pytest.raises(ValueError, match="unknown pseudoword generation method"):
        build_lexdec_stimuli(pool, 10, reference_words=lex["word"].tolist(), method="invalid")

File: R_workflow/tests/testthat/test-generation-subsyllabic.R
test_that("subsyllabic segmentation matches worked examples", {
  expect_equal(segment_subsyllabic("bridge"),
               list(list(role = "onset", text = "br"), list(role = "nucleus", text = "i"),
                    list(role = "coda", text = "d"), list(role = "onset", text = "g"),
                    list(role = "nucleus", text = "e")))
  expect_equal(segment_subsyllabic("strength"),
               list(list(role = "onset", text = "str"), list(role = "nucleus", text = "e"),
                    list(role = "coda", text = "ngth")))
  expect_length(segment_subsyllabic("crwth"), 0)          # no vowel run
})

test_that("subsyllabic generation yields legal non-words of preserved length", {
  schema <- yaml::read_yaml(system.file("extdata", "schema.yaml", package = "lexsync"))
  lex <- load_lexicon(system.file("extdata", "en_example.csv", package = "lexsync"), schema, "english")
  ref <- lex$word
  base <- head(build_pool(lex, list(length = c(4, 7)))$word, 30)
  gen <- generate_pseudowords_subsyllabic(base, ref)
  bg <- bigram_counts(ref)
  for (i in seq_len(nrow(gen))) {
    expect_equal(nchar(gen$pseudoword[i]), nchar(gen$base_word[i]))   # length preserved
    expect_false(gen$pseudoword[i] %in% ref)                          # a novel non-word
    expect_true(.bg_legal(gen$pseudoword[i], bg))                     # every bigram attested
  }
  expect_equal(length(unique(gen$pseudoword)), nrow(gen))
})

test_that("build_lexdec_stimuli rejects an unknown generation method", {
  schema <- yaml::read_yaml(system.file("extdata", "schema.yaml", package = "lexsync"))
  lex <- load_lexicon(system.file("extdata", "en_example.csv", package = "lexsync"), schema, "english")
  pool <- build_pool(lex, list(length = c(4, 6)))
  # The set is named, as every other 'unknown X' message in the package names it;
  # test_generation_subsyllabic.py pins the same sentence.
  expect_error(build_lexdec_stimuli(pool, 10L, reference_words = lex$word, method = "invalid"),
               paste("lexsync: unknown pseudoword generation method 'invalid'.",
                     "Known methods: letter_substitution, subsyllabic."),
               fixed = TRUE)
})
